code/controllers/subsystems/explosives.dm

```diff
@@ -212,78 +216,190 @@ var/datum/controller/subsystem/explosives/SSexplosives
-	for(var/direction in cardinal)
-		var/turf/T = get_step(epicenter, direction)
-		if (T)
-			explosion_spread(T, power - epicenter.explosion_resistance, direction)
+		if (!T)
 			CHECK_TICK
+			continue
 
-	//This step applies the ex_act effects for the explosion, as planned in the previous step.
-	for(var/turf/T in explosion_turfs)
-		if(!T || explosion_turfs[T] <= 0)
+		if (!ARE_Z_CONNECTED(T.z, epicenter.z))
+			CHECK_TICK
+			continue
+
+		if (T.type == /turf/space)	// Equality is faster than istype.
+			reception = EXPLFX_NONE
+
+			for (var/turf/simulated/THING in RANGE_TURFS(1, M))
+				reception |= EXPLFX_SHAKE
+				break
+
+			if (!reception)
+				CHECK_TICK
+				continue
+
+		var/dist = get_dist(M, epicenter) || 1
+		if ((reception & EXPLFX_SOUND) && M.ear_deaf <= 0)
+			if (dist <= close_dist)
+				M.playsound_local(epicenter, explosion_sound, min(100, volume), 1, frequency, falloff = 5)
+				//You hear a far explosion if you're outside the blast radius. Small bombs shouldn't be heard all over the station.
+			else
+				volume = M.playsound_local(epicenter, 'sound/effects/explosionfar.ogg', volume, 1, frequency, usepressure = 0, falloff = 1000)
+
+		if ((reception & EXPLFX_SHAKE) && volume > 0)
+			shake_camera(M, min(30, max(2,(power*2) / dist)), min(3.5, ((power/3) / dist)),0.05)
+			//Maximum duration is 3 seconds, and max strength is 3.5
+			//Becuse values higher than those just get really silly
+
+		CHECK_TICK
+
+	log_debug("iexpl: SFX phase completed in [(world.time-time)/10] seconds.")
+	log_debug("iexpl: Beginning application phase.")
+	time = world.time
+
+	var/turf_tally = 0
+	var/movable_tally = 0
+	for (var/thing in act_turfs)
+		var/turf/T = thing
+		if (act_turfs[T] <= 0)
 			CHECK_TICK
 			continue
 
 		//Wow severity looks confusing to calculate... Fret not, I didn't leave you with any additional instructions or help. (just kidding, see the line under the calculation)
-		var/severity = 4 - round(max(min( 3, ((explosion_turfs[T] - T.explosion_resistance) / (max(3,(power/3)))) ) ,1), 1)								//sanity			effective power on tile				divided by either 3 or one third the total explosion power
-								//															One third because there are three power levels and I
-								//															want each one to take up a third of the crater
-		var/x = T.x
-		var/y = T.y
-		var/z = T.z
-		T.ex_act(severity)
-		if(!T)
-			T = locate(x,y,z)
-		for(var/atom/A in T)
-			A.ex_act(severity)
+		var/severity = 4 - round(max(min( 3, ((act_turfs[T] - T.explosion_resistance) / (max(3,(power/3)))) ) ,1), 1)
+		//sanity			effective power on tile				divided by either 3 or one third the total explosion power
+		//															One third because there are three power levels and I
+		//															want each one to take up a third of the crater
+
+		if (T.simulated)
+			INVOKE_ASYNC(T, /atom/.proc/ex_act, severity)
+			T.ex_act(severity)
```

Comment:
:thinking:

Reply:
Lohikar are you reviewing your own code.